### ISSUE TITLE
Add ability to disable verification

### ISSUE THREAD

**Issue by seanmonstar** [sfackler/rust-native-tls#13]
I know, I know, it's horrible. It is also a reality that many people deal with this, and so it would be useful to add a way to disable verification. It can and probably should be a terrifying long name, `very_insecure_disable_verification` or what-have-you.

cc https://github.com/seanmonstar/reqwest/issues/15

**Comment by sfackler** [comment]
How much of verification? Just hostname or the whole deal?

**Comment by seanmonstar** [comment]
I assume certificate verification. Besides the host name verification, the rest is to verify from a set of trusted CAs that the certificate is signed by one of them? I've never needed it myself, but the case I imagine is people have a self-signed certificate. In that case, hostname still makes sense, but checking the CAs would cause problems.

**Comment by alex** [comment]
I'd start with making it easier to specify an _alternate_ validator, e.g. provide a `PinnedValidator` which people can put their self-signed cert in to trust that. Try to find ways to support every semi-reasonable usecase that people want to disable stuff for. And only when that finally fails add the degenerate `VeryInsecureNoValidationValidator`

**Comment by sfackler** [comment]
Definitely agree on that approach.

Adding trusted CAs is in some ways orthogonal to the validators, and is just blocked on adding support to schannel-rs. Unfortunately, this is not really a thing Windows wants you to do (trusted CAs must either be registered in the system trust root, or be present in a Certificate Trust List signed by something trusted by your system trust root) but it should be doable.

**Comment by Philipp91** [comment]
Maybe the `VeryInsecureNoValidationValidator` could be hidden behind a cargo feature called `make_me_vulnerable_to_mitm` or something like that. The concrete security implications should be made clear -- not only the general fact that it's less secure.

**Comment by sfackler** [comment]
Hostname verification is in #15.

**Comment by emk** [comment]
I'm currently dealing with a production emergency involving `native-tls`, PostgreSQL and Citus's hosted database product. This is occurring in my company's first production Rust app. :-( (Happily, nobody's trying to use it this afternoon, but I need to fix it quickly nonetheless.)

The problem is that PostgreSQL SSL cert validation is still fairly rare in non-Rust libraries, and plenty of PostgreSQL servers have self-signed or invalidly-signed certs. It's hard to debug because PostgreSQL uses [a custom protocol to enable SSL](https://www.postgresql.org/docs/9.3/static/protocol-flow.html#AEN99579). I'd be debugging faster if I weren't sleep deprived from work travel. :-(

Because I chose `native-tls`, there's no easy way for me to disable cert validation for the duration of the production emergency.

So :+1: for making this option, even if it's has an obnoxiously long and scary name.

**Comment by vignesh-sankaran** [comment]
I've run into this problem too. Currently am writing integration tests for an API, and I use self signed certs in the dev environment.

**Comment by bfrog** [comment]
I just ran into this today, updated a bunch of stuff and switch to native-tls from openssl for postgres, got an error verifying the server certs. Is there at least a workaround for disabling verification?

**Comment by devunt** [comment]
Will definitely great when working on local connection with self-signed certificates. Requiring valid certificate issued from trusted CA would be too harsh for 127.0.0.1.

**Comment by sfackler** [comment]
Self signed dev certificates should be taken care of by https://github.com/sfackler/rust-native-tls/pull/28

**Comment by Boscop** [comment]
I also need the ability to accept invalid ssl certs at my work..

**Comment by glyn** [comment]
@sfackler What's the simplest way to handle a self-signed dev certificate by #28? It really needs to be simple 'cos it's common to set up test servers using self-signed certs. :-)

**Comment by sfackler** [comment]
Build a `Certificate` from the DER-encoded self signed cert and pass it to `TlsConnectorBuilder::add_root_certificate`.

**Comment by glyn** [comment]
Hmmm. I personally think that's too complex, especially in situations when there are a whole host of servers with different self-signed certs. It would be much simpler and more usable to have an "ignore SSL validation" type option with suitable caveats.

**Comment by yoni386** [comment]
is there any way to disable ssl validation for self sign certificate? name validation? open connection to ip by name

**Comment by sfackler** [comment]
@yoni386 you can connect without name validation: https://docs.rs/native-tls/0.1.2/native_tls/struct.TlsConnector.html#method.danger_connect_without_providing_domain_for_certificate_verification_and_server_name_indication.

**Comment by yoni386** [comment]
This now works 216.58.205.238 (google.com)

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Failure(Ssl(ErrorStack([Error { code: 336134278, library: "SSL routines", function: "SSL3_GET_SERVER_CERTIFICATE", reason: "certificate verify failed", file: "s3_clnt.c", line: 1178 }])))', /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/result.rs:868
stack backtrace:
   1:     0x56232b05ae8c - std::sys::imp::backtrace::tracing::imp::write::hf33ae72d0baa11ed
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x56232b05cfbe - std::panicking::default_hook::{{closure}}::h59672b733cc6a455
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:351
   3:     0x56232b05cbc4 - std::panicking::default_hook::h1670459d2f3f8843
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:367
   4:     0x56232b05d45b - std::panicking::rust_panic_with_hook::hcf0ddb069e7beee7
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:555
   5:     0x56232b05d2f4 - std::panicking::begin_panic::hd6eb68e27bdf6140
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:517
   6:     0x56232b05d219 - std::panicking::begin_panic_fmt::hfea5965948b877f8
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:501
   7:     0x56232b05d1a7 - rust_begin_unwind
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:477
   8:     0x56232b0835cd - core::panicking::panic_fmt::hc0f6d7b2c300cdd9
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/panicking.rs:69
   9:     0x56232b03b783 - core::result::unwrap_failed::h5b121a371498c6c5
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/macros.rs:29
  10:     0x56232b039367 - <core::result::Result<T, E>>::unwrap::h73c9898ca06e4a16
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/result.rs:745
  11:     0x56232b0418cb - httpsc::main::h9383f77d63f89ac5
                        at /opt/dev/httpsc/src/main.rs:15
  12:     0x56232b063f6a - __rust_maybe_catch_panic
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libpanic_unwind/lib.rs:98
  13:     0x56232b05d966 - std::rt::lang_start::hd7c880a37a646e81
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:436
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panic.rs:361
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/rt.rs:57
  14:     0x56232b041c32 - main
  15:     0x7f64aa122d1c - __libc_start_main
  16:     0x56232b0332d8 - <unknown>

**Comment by mark-blum** [comment]
Hello, I have a use case at work where a development server has both a self signed and an expired certificate. It would be great if I could use the certificate even though it is expired.

**Comment by algermissen** [comment]
I'd like to add this here for reference in case someone finds it helpful: http://www.jalg.net/2017/09/hyper-client-and-self-signed-certs

**Comment by H2CO3** [comment]
Hi there @sfackler,

I'm experiencing a problem with certificate pinning. I'm trying to connect to an HTTPS server running on my machine. I've set up the server using Hyper and the TLS server from `hyper-native-tls`. I can successfully connect to the server from an iOS app in which I've implemented public key pinning. I can also connect to the server using `curl --insecure`. Following the few related issues on GitHub, I'm trying to add my self-signed server certificate as trusted using `native-tls`, by doing this:

```rust
use std::io::Read;
use std::fs::File;

let mut cert_data = Vec::new();
let mut file = File::open("self_signed_cert.der")?;
file.read_to_end(&mut cert_data)?;

let cert = Certificate::from_der(&cert_data)?;
let mut tls_connector_builder = TlsConnector::builder()?;
tls_connector_builder.add_root_certificate(cert)?;
let tls_connector = tls_connector_builder.build()?;
let tls_client = NativeTlsClient::from(tls_connector);
let https_connector = HttpsConnector::new(tls_client);
let https_client = Client::with_connector(https_connector);

https_client
    .post("https://localhost:3598/")
    .send()
    .expect("error response");
```

However, the request fails with the not very helpful message:

```
thread 'main' panicked at 'error response: Underlying(Ssl(Failure(Error { code: -67843, message: "The trust policy was not trusted." })))', src/libcore/result.rs:906:4
```

Adding my server's own certificate to the client should have solved this problem, shouldn't it? What may I be doing wrong?

**Comment by algermissen** [comment]
Hi @H2CO3 , not sure if it helps, but this worked for me: http://www.jalg.net/2017/09/hyper-client-and-self-signed-certs

Maybe you also need to disable verification as discussed here https://github.com/sfackler/rust-native-tls/issues/13#issuecomment-278448563

**Comment by sfackler** [comment]
@H2CO3 did you add the root certificate or a leaf?

**Comment by H2CO3** [comment]
Hey @algermissen, unfortunately I'm using the Iron framework which depends on version 0.10 of Hyper, so that solution (which requires Hyper 0.11) doesn't work for me. :(

By disabling verification, do you mean disabling **hostname** verification? (It doesn't look like anything else is implemented, based on the discussion on that pull request, just wanted to be sure.)

@sfackler hey, thanks for the quick reply! I only have one, self-signed certificate. I'm not sure if it qualifies as a root or leaf…

**Comment by sfackler** [comment]
Things should be working then. We should be running this test on iOS in CI - can you try it out on your end as well? https://github.com/sfackler/rust-native-tls/blob/master/src/test.rs#L56

**Comment by H2CO3** [comment]
Tests pass on my machine. As far as I can tell, the only thing that can differ in the test function you linked to and my code is the server side. In my project I'm using Iron and hyper-native-tls like this:

```rust
    let tls = NativeTlsServer::new(&config.tls_cert_file, &config.tls_cert_pwd)?;

    Iron::new(global_context_handler)
        .https(listen_addr, tls)
        .map_err(Into::into)
```

Which works and the server starts, so the cert file and its password, I assume, must be correct as well… very strange.

**Comment by sfackler** [comment]
Hmm, what happens if you try to run the rust-native-tls test but using your certificate/key pair rather than the ones it's currently using?

**Comment by H2CO3** [comment]
If I plug my exact same `.p12`/password and `.der` combination into the rust-native-tls tests, then tests pass, even if I set the certificate explicitly untrusted in my keychain. So that works correctly.

However, the same code with the same files doesn't seem to work when my client with the same code tries to connect to my server. o.O

**Comment by H2CO3** [comment]
Okay, when I changed the hostname in the test, it started failing. This indicated to me that hostname verification was still active. I disabled it in my client and voila! it worked.

I now only have to figure out how to tell the `hyper-native-tls` client to verify against a particular hostname when connecting to `127.0.0.1`.

Thanks for your help!

**Comment by sfackler** [comment]
Implemented by #56
